File: bindings/rust/idax/src/navigation.rs
//! Opaque persistent address navigation history.

use std::ffi::{CStr, CString, c_char, c_void};

use crate::address::Address;
use crate::error::{self, Error, Result, Status};

/// Owned snapshot of one semantic navigation location.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Entry {
    pub address: Address,
    pub channel: String,
    pub metadata: String,
}

impl Entry {
    #[must_use]
    pub fn new(address: Address, channel: impl Into<String>, metadata: impl Into<String>) -> Self {
        Self {
            address,
            channel: channel.into(),
            metadata: metadata.into(),
        }
    }
}

struct RawEntry {
    raw: idax_sys::IdaxNavigationEntry,
    _channel: CString,
    _metadata: CString,
}

fn cstring(value: &str, field: &str) -> Result<CString> {
    CString::new(value)
        .map_err(|_| Error::validation(format!("{field} contains an embedded NUL byte")))
}

fn raw_entry(entry: &Entry) -> Result<RawEntry> {
    let channel = cstring(&entry.channel, "navigation channel")?;
    let metadata = cstring(&entry.metadata, "navigation metadata")?;
    let raw = idax_sys::IdaxNavigationEntry {
        address: entry.address,
        channel: channel.as_ptr().cast_mut(),
        metadata: metadata.as_ptr().cast_mut(),
    };
    Ok(RawEntry {
        raw,
        _channel: channel,
        _metadata: metadata,
    })
}

fn empty_raw_entry() -> idax_sys::IdaxNavigationEntry {
    idax_sys::IdaxNavigationEntry {
        address: 0,
        channel: std::ptr::null_mut(),
        metadata: std::ptr::null_mut(),
    }
}

unsafe fn copy_raw_entry(raw: &idax_sys::IdaxNavigationEntry) -> Result<Entry> {
    if raw.channel.is_null() || raw.metadata.is_null() {
        return Err(Error::internal(
            "navigation entry contains a null string pointer",
        ));
    }
    let channel = unsafe { CStr::from_ptr(raw.channel) }
        .to_str()
        .map_err(|_| Error::internal("navigation channel is not valid UTF-8"))?
        .to_owned();
    let metadata = unsafe { CStr::from_ptr(raw.metadata) }
        .to_str()
        .map_err(|_| Error::internal("navigation metadata is not valid UTF-8"))?
        .to_owned();
    Ok(Entry {
        address: raw.address,
        channel,
        metadata,
    })
}

fn entry_result(
    call: impl FnOnce(*mut idax_sys::IdaxNavigationEntry) -> i32,
    fallback: &str,
) -> Result<Entry> {
    let mut raw = empty_raw_entry();
    let status = call(&mut raw);
    if status != 0 {
        unsafe { idax_sys::idax_navigation_entry_free(&mut raw) };
        return Err(error::consume_last_error(fallback));
    }
    let result = unsafe { copy_raw_entry(&raw) };
    unsafe { idax_sys::idax_navigation_entry_free(&mut raw) };
    result
}

fn optional_entry_result(
    call: impl FnOnce(*mut idax_sys::IdaxNavigationEntry, *mut i32) -> i32,
    fallback: &str,
) -> Result<Option<Entry>> {
    let mut raw = empty_raw_entry();
    let mut has_value = 0;
    let status = call(&mut raw, &mut has_value);
    if status != 0 {
        unsafe { idax_sys::idax_navigation_entry_free(&mut raw) };
        return Err(error::consume_last_error(fallback));
    }
    if has_value == 0 {
        return Ok(None);
    }
    let result = unsafe { copy_raw_entry(&raw) };
    unsafe { idax_sys::idax_navigation_entry_free(&mut raw) };
    result.map(Some)
}

fn entries_result(
    call: impl FnOnce(*mut *mut idax_sys::IdaxNavigationEntry, *mut usize) -> i32,
    fallback: &str,
) -> Result<Vec<Entry>> {
    let mut raw = std::ptr::null_mut();
    let mut count = 0;
    let status = call(&mut raw, &mut count);
    if status != 0 {
        unsafe { idax_sys::idax_navigation_entries_free(raw, count) };
        return Err(error::consume_last_error(fallback));
    }
    if count != 0 && raw.is_null() {
        return Err(Error::internal(
            "navigation entry array pointer is null for a nonempty result",
        ));
    }
    let mut result = Vec::with_capacity(count);
    if !raw.is_null() {
        let values = unsafe { std::slice::from_raw_parts(raw, count) };
        for value in values {
            match unsafe { copy_raw_entry(value) } {
                Ok(value) => result.push(value),
                Err(failure) => {
                    unsafe { idax_sys::idax_navigation_entries_free(raw, count) };
                    return Err(failure);
                }
            }
        }
        unsafe { idax_sys::idax_navigation_entries_free(raw, count) };
    }
    Ok(result)
}

/// Owned opaque handle to one persistent, IDAX-private history stream.
#[derive(Debug)]
pub struct History {
    handle: *mut c_void,
}

impl History {
    /// Open or create a logical stream with one initial tip.
    pub fn open(name: &str, initial: &Entry) -> Result<Self> {
        let name = cstring(name, "navigation history name")?;
        let initial = raw_entry(initial)?;
        let mut handle = std::ptr::null_mut();
        let status = unsafe {
            idax_sys::idax_navigation_history_open(name.as_ptr(), &initial.raw, &mut handle)
        };
        if status != 0 {
            return Err(error::consume_last_error(
                "navigation::History::open failed",
            ));
        }
        if handle.is_null() {
            return Err(Error::internal(
                "navigation history open returned a null handle",
            ));
        }
        Ok(Self { handle })
    }

    /// Return the caller-visible logical stream name.
    pub fn name(&self) -> Result<String> {
        let mut out: *mut c_char = std::ptr::null_mut();
        let status = unsafe { idax_sys::idax_navigation_history_name(self.handle, &mut out) };
        if status != 0 {
            return Err(error::consume_last_error(
                "navigation::History::name failed",
            ));
        }
        unsafe { error::cstr_to_string_free(out, "navigation history name is null") }
    }

    /// Report whether this handle's open call created the stream.
    pub fn created(&self) -> Result<bool> {
        let mut out = 0;
        let status = unsafe { idax_sys::idax_navigation_history_created(self.handle, &mut out) };
        if status != 0 {
            return Err(error::consume_last_error(
                "navigation::History::created failed",
            ));
        }
        Ok(out != 0)
    }

    /// Copy every stack entry in index order.
    pub fn entries(&self) -> Result<Vec<Entry>> {
        entries_result(
            |out, count| unsafe {
                idax_sys::idax_navigation_history_entries(self.handle, out, count)
            },
            "navigation::History::entries failed",
        )
    }

    /// Copy the number of stack entries.
    pub fn size(&self) -> Result<usize> {
        let mut out = 0;
        let status = unsafe { idax_sys::idax_navigation_history_size(self.handle, &mut out) };
        if status != 0 {
            return Err(error::consume_last_error(
                "navigation::History::size failed",
            ));
        }
        Ok(out)
    }

    /// Copy the current stack index.
    pub fn index(&self) -> Result<usize> {
        let mut out = 0;
        let status = unsafe { idax_sys::idax_navigation_history_index(self.handle, &mut out) };
        if status != 0 {
            return Err(error::consume_last_error(
                "navigation::History::index failed",
            ));
        }
        Ok(out)
    }

    /// Copy the entry at the current stack index.
    pub fn current(&self) -> Result<Entry> {
        entry_result(
            |out| unsafe { idax_sys::idax_navigation_history_current(self.handle, out) },
            "navigation::History::current failed",
        )
    }

    /// Copy one channel's current location, or absence.
    pub fn current_for(&self, channel: &str) -> Result<Option<Entry>> {
        let channel = cstring(channel, "navigation channel")?;
        optional_entry_result(
            |out, has_value| unsafe {
                idax_sys::idax_navigation_history_current_for(
                    self.handle,
                    channel.as_ptr(),
                    out,
                    has_value,
                )
            },
            "navigation::History::current_for failed",
        )
    }

    /// Copy every channel-current location; ordering is host-defined.
    pub fn all_current(&self) -> Result<Vec<Entry>> {
        entries_result(
            |out, count| unsafe {
                idax_sys::idax_navigation_history_all_current(self.handle, out, count)
            },
            "navigation::History::all_current failed",
        )
    }

    /// Update one channel's current location, optionally replacing the cursor.
    pub fn set_current(&self, entry: &Entry, record_in_history: bool) -> Status {
        let entry = raw_entry(entry)?;
        let status = unsafe {
            idax_sys::idax_navigation_history_set_current(
                self.handle,
                &entry.raw,
                record_in_history as i32,
            )
        };
        error::int_to_status(status, "navigation::History::set_current failed")
    }

    /// Append after the cursor, truncating forward entries.
    pub fn push(&self, entry: &Entry) -> Result<Entry> {
        let entry = raw_entry(entry)?;
        entry_result(
            |out| unsafe { idax_sys::idax_navigation_history_push(self.handle, &entry.raw, out) },
            "navigation::History::push failed",
        )
    }

    /// Move the cursor to an exact existing index.
    pub fn seek(&self, index: usize) -> Result<Entry> {
        entry_result(
            |out| unsafe { idax_sys::idax_navigation_history_seek(self.handle, index, out) },
            "navigation::History::seek failed",
        )
    }

    /// Move backward, or return absence at the boundary.
    pub fn back(&self, count: usize) -> Result<Option<Entry>> {
        optional_entry_result(
            |out, has_value| unsafe {
                idax_sys::idax_navigation_history_back(self.handle, count, out, has_value)
            },
            "navigation::History::back failed",
        )
    }

    /// Move forward, or return absence at the boundary.
    pub fn forward(&self, count: usize) -> Result<Option<Entry>> {
        optional_entry_result(
            |out, has_value| unsafe {
                idax_sys::idax_navigation_history_forward(self.handle, count, out, has_value)
            },
            "navigation::History::forward failed",
        )
    }

    /// Replace one indexed entry without changing size or cursor.
    pub fn replace(&self, index: usize, entry: &Entry) -> Status {
        let entry = raw_entry(entry)?;
        let status =
            unsafe { idax_sys::idax_navigation_history_replace(self.handle, index, &entry.raw) };
        error::int_to_status(status, "navigation::History::replace failed")
    }

    /// Replace the complete stack with one tip at index zero.
    pub fn clear(&self, new_tip: &Entry) -> Status {
        let new_tip = raw_entry(new_tip)?;
        let status = unsafe { idax_sys::idax_navigation_history_clear(self.handle, &new_tip.raw) };
        error::int_to_status(status, "navigation::History::clear failed")
    }

    /// Move one channel to another distinct history.
    pub fn transfer_channel_to(
        &self,
        destination: &History,
        channel: &str,
        retain_history: bool,
    ) -> Status {
        let channel = cstring(channel, "navigation channel")?;
        let status = unsafe {
            idax_sys::idax_navigation_history_transfer_channel_to(
                self.handle,
                destination.handle,
                channel.as_ptr(),
                retain_history as i32,
            )
        };
        error::int_to_status(status, "navigation::History::transfer_channel_to failed")
    }
}

impl Drop for History {
    fn drop(&mut self) {
        unsafe { idax_sys::idax_navigation_history_free(self.handle) };
        self.handle = std::ptr::null_mut();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn entry_constructor_preserves_owned_state() {
        let entry = Entry::new(0x401000, "disassembly", "cursor");
        assert_eq!(entry.address, 0x401000);
        assert_eq!(entry.channel, "disassembly");
        assert_eq!(entry.metadata, "cursor");
    }

    #[test]
    fn embedded_nul_is_rejected_before_ffi() {
        let invalid_channel = Entry::new(0x401000, "bad\0channel", "metadata");
        assert!(raw_entry(&invalid_channel).is_err());
        let invalid_metadata = Entry::new(0x401000, "channel", "bad\0metadata");
        assert!(raw_entry(&invalid_metadata).is_err());
    }

    #[test]
    fn signatures_preserve_owned_and_optional_state() {
        let _: fn(&str, &Entry) -> Result<History> = History::open;
        let _: fn(&History) -> Result<Vec<Entry>> = History::entries;
        let _: fn(&History, &str) -> Result<Option<Entry>> = History::current_for;
        let _: fn(&History, usize) -> Result<Option<Entry>> = History::back;
        let _: fn(&History, usize) -> Result<Option<Entry>> = History::forward;
        let _: fn(&History, &History, &str, bool) -> Status = History::transfer_channel_to;
    }
}

File: bindings/swift/Sources/IDAX/Navigation.swift
internal import CIDAX

extension Navigation {
  public struct Entry: Equatable, Sendable {
    public var address: Address
    public var channel: String
    public var metadata: String
    public init(address: Address = badAddress, channel: String = "", metadata: String = "") {
      self.address = address
      self.channel = channel
      self.metadata = metadata
    }
    internal init(_ value: IdaxNavigationEntry, _ operation: String) throws(IDAError) {
      address = value.address
      channel = try borrowCString(value.channel.map { UnsafePointer($0) }, operation)
      metadata = try borrowCString(value.metadata.map { UnsafePointer($0) }, operation)
    }
    fileprivate func native<Value>(
      _ operation: String,
      _ body: (UnsafePointer<IdaxNavigationEntry>) -> Value
    ) throws(IDAError) -> Value {
      try checkedCStringArray([channel, metadata], operation) { strings, _ in
        var value = IdaxNavigationEntry(
          address: address, channel: UnsafeMutablePointer(mutating: strings![0]),
          metadata: UnsafeMutablePointer(mutating: strings![1]))
        return withUnsafePointer(to: &value, body)
      }
    }
  }

  /// A persistent history stream, confined to the database session that opened it.
  public final class History {
    private let resource: NativeResource
    public let name: String
    public let created: Bool

    private init(taking pointer: UnsafeMutableRawPointer?, _ operation: String) throws(IDAError) {
      guard let pointer else {
        throw IDAError(
          category: .internalError, message: "Native history is null", context: operation)
      }
      resource = try NativeResource(
        taking: pointer, release: idax_navigation_history_free, operation: operation)
      name = try withStringOutput(operation) { idax_navigation_history_name(pointer, $0) }
      created =
        try withOutput(operation, initial: Int32(0)) {
          idax_navigation_history_created(pointer, $0)
        } != 0
    }
    public static func open(_ name: String, initial: Entry) throws(IDAError) -> History {
      let operation = "Navigation.History.open"
      try requireRuntimeThread(operation)
      var output: UnsafeMutableRawPointer?
      try checkStatus(
        checkedCStringArray([name, initial.channel, initial.metadata], operation) { strings, _ in
          var entry = IdaxNavigationEntry(
            address: initial.address,
            channel: UnsafeMutablePointer(mutating: strings![1]),
            metadata: UnsafeMutablePointer(mutating: strings![2]))
          return idax_navigation_history_open(strings![0], &entry, &output)
        }, operation)
      return try History(taking: output, operation)
    }
    public func close() throws(IDAError) { try resource.close("Navigation.History.close") }
    public func copy() throws(IDAError) -> History {
      let operation = "Navigation.History.copy"
      let pointer = try resource.pointer(operation)
      var output: UnsafeMutableRawPointer?
      try bridgeCall(operation) { idax_swift_navigation_clone(pointer, &output, $0) }
      return try History(taking: output, operation)
    }

    private func entries(
      _ operation: String,
      _ body: (
        UnsafeMutableRawPointer?, UnsafeMutablePointer<UnsafeMutablePointer<IdaxNavigationEntry>?>,
        UnsafeMutablePointer<Int>
      ) -> Int32
    ) throws(IDAError) -> [Entry] {
      let pointer = try resource.pointer(operation)
      var output: UnsafeMutablePointer<IdaxNavigationEntry>?
      var count = 0
      defer { idax_navigation_entries_free(output, count) }
      try checkStatus(body(pointer, &output, &count), operation)
      let buffer = try checkedBuffer(output.map { UnsafePointer($0) }, count: count, operation)
      var result: [Entry] = []
      result.reserveCapacity(count)
      for entry in buffer { result.append(try Entry(entry, operation)) }
      return result
    }
    public func entries() throws(IDAError) -> [Entry] {
      try entries("Navigation.History.entries", idax_navigation_history_entries)
    }
    public func allCurrent() throws(IDAError) -> [Entry] {
      try entries("Navigation.History.allCurrent", idax_navigation_history_all_current)
    }
    public func size() throws(IDAError) -> Int {
      let pointer = try resource.pointer("Navigation.History.size")
      return try withOutput("Navigation.History.size", initial: 0) {
        idax_navigation_history_size(pointer, $0)
      }
    }
    public func index() throws(IDAError) -> Int {
      let pointer = try resource.pointer("Navigation.History.index")
      return try withOutput("Navigation.History.index", initial: 0) {
        idax_navigation_history_index(pointer, $0)
      }
    }
    public func current() throws(IDAError) -> Entry {
      let operation = "Navigation.History.current"
      let pointer = try resource.pointer(operation)
      var output = IdaxNavigationEntry()
      defer { idax_navigation_entry_free(&output) }
      try checkStatus(idax_navigation_history_current(pointer, &output), operation)
      return try Entry(output, operation)
    }
    public func current(for channel: String) throws(IDAError) -> Entry? {
      let operation = "Navigation.History.current(for:)"
      let pointer = try resource.pointer(operation)
      var output = IdaxNavigationEntry()
      var present: Int32 = 0
      defer { idax_navigation_entry_free(&output) }
      try checkStatus(
        checkedCString(channel, operation) {
          idax_navigation_history_current_for(pointer, $0, &output, &present)
        }, operation)
      return present == 0 ? nil : try Entry(output, operation)
    }
    public func setCurrent(_ entry: Entry, recordInHistory: Bool = false) throws(IDAError) {
      let operation = "Navigation.History.setCurrent"
      let pointer = try resource.pointer(operation)
      try checkStatus(
        entry.native(operation) {
          idax_navigation_history_set_current(pointer, $0, recordInHistory ? 1 : 0)
        }, operation)
    }
    @discardableResult public func push(_ entry: Entry) throws(IDAError) -> Entry {
      let operation = "Navigation.History.push"
      let pointer = try resource.pointer(operation)
      var output = IdaxNavigationEntry()
      defer { idax_navigation_entry_free(&output) }
      try checkStatus(
        entry.native(operation) { idax_navigation_history_push(pointer, $0, &output) }, operation)
      return try Entry(output, operation)
    }
    @discardableResult public func seek(to index: Int) throws(IDAError) -> Entry {
      let operation = "Navigation.History.seek"
      guard index >= 0 else {
        throw IDAError(
          category: .validation, message: "History index cannot be negative", context: operation)
      }
      let pointer = try resource.pointer(operation)
      var output = IdaxNavigationEntry()
      defer { idax_navigation_entry_free(&output) }
      try checkStatus(idax_navigation_history_seek(pointer, index, &output), operation)
      return try Entry(output, operation)
    }
    private func move(
      _ count: Int, _ operation: String,
      _ body: (
        UnsafeMutableRawPointer?, Int, UnsafeMutablePointer<IdaxNavigationEntry>,
        UnsafeMutablePointer<Int32>
      ) -> Int32
    ) throws(IDAError) -> Entry? {
      guard count >= 0 else {
        throw IDAError(
          category: .validation, message: "History movement count cannot be negative",
          context: operation)
      }
      let pointer = try resource.pointer(operation)
      var output = IdaxNavigationEntry()
      var present: Int32 = 0
      defer { idax_navigation_entry_free(&output) }
      try checkStatus(body(pointer, count, &output, &present), operation)
      return present == 0 ? nil : try Entry(output, operation)
    }
    @discardableResult public func back(_ count: Int = 1) throws(IDAError) -> Entry? {
      try move(count, "Navigation.History.back", idax_navigation_history_back)
    }
    @discardableResult public func forward(_ count: Int = 1) throws(IDAError) -> Entry? {
      try move(count, "Navigation.History.forward", idax_navigation_history_forward)
    }
    public func replace(at index: Int, with entry: Entry) throws(IDAError) {
      let operation = "Navigation.History.replace"
      guard index >= 0 else {
        throw IDAError(
          category: .validation, message: "History index cannot be negative", context: operation)
      }
      let pointer = try resource.pointer(operation)
      try checkStatus(
        entry.native(operation) { idax_navigation_history_replace(pointer, index, $0) }, operation)
    }
    public func clear(newTip: Entry) throws(IDAError) {
      let operation = "Navigation.History.clear"
      let pointer = try resource.pointer(operation)
      try checkStatus(
        newTip.native(operation) { idax_navigation_history_clear(pointer, $0) }, operation)
    }
    public func transferChannel(
      _ channel: String, to destination: History, retainHistory: Bool = true
    ) throws(IDAError) {
      let operation = "Navigation.History.transferChannel"
      let pointer = try resource.pointer(operation)
      let target = try destination.resource.pointer(operation)
      try checkStatus(
        checkedCString(channel, operation) {
          idax_navigation_history_transfer_channel_to(pointer, target, $0, retainHistory ? 1 : 0)
        }, operation)
    }
  }
}
